---
title: "scRNA-seq Pipeline"
author: "Emily Nissen"
date: "2024-02-07"
output: 
  github_document:
    html_preview: TRUE
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Project Notes

Add any project notes here

# Pre-processing

See the "PreProcess/" folder for scripts that can be downloaded based on the code in the following steps.

These steps are specifically for data from 10x genomics.

## 0_Cell_Ranger_Mkfastq

Demultiplex - if needed. Most likely you do not need to demultiplex yourself. With the new NovaSeq X Plus Sequencer, the genomics core is able to perform the bcl2fastq on instrument for 10x single-cell data.

```{bash, eval = F}
/path/to/cellranger mkfastq --id=fastq \
--run=/path/to/run/folder \
--samplesheet=/path/to/run/folder/SampleSheet-1.csv
```

## 0_Cell_Ranger

Run cell ranger count. All you need to supply is how many jobs you want to create (by setting the number of arrays) and the sample sheet file from the run. The "Sample_ID" column in this sample sheet file should match the names of the folders that contain the fastq files


# Analysis

https://hbctraining.github.io/scRNA-seq/lessons/04_SC_quality_control.html

```{r setup2, include=FALSE, eval = F}
knitr::opts_chunk$set(echo = TRUE, fig.height = 8, fig.width = 8)
library(Seurat)
# library(EnsDb.Hsapiens.v86)
# library(BSgenome.Hsapiens.UCSC.hg38)
library(MASS)
library(ggplot2)
library(patchwork)
library(tidyr)
library(dplyr)
library(clusterProfiler)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(ReactomePA)
library(rio)
library(scDblFinder)
library(SingleCellExperiment)
library(BiocParallel)
library(PCAtools)
library(viridis)
library(RColorBrewer)
library(ggpubr)
library(DirichletReg)
library(edgeR)

data.path = "/path/to/folders/output/from/cellranger"
robjects.path = "/path/to/robjects/folder"
tables.path = "/path/to/tables/folder"
```

## Create Seurat Objects

```{r create, eval = F}
samples = list.files(data.path)
## set up function
read.data <- function(sample){
  counts = Read10X_h5(paste0(data.path, sample, "/outs/filtered_feature_bc_matrix.h5"))


  # create a Seurat object containing the RNA data
  obj = CreateSeuratObject(
    counts = counts,
    assay = "RNA"
  )

  obj[["percent.mt"]] = PercentageFeatureSet(obj, pattern = "^MT-")
  obj[["percent.ribosomal"]] = PercentageFeatureSet(obj, pattern = "^RP[LS]")

  return(obj)
}

seurat.list = list()

for (i in 1:length(samples)){
  sample <- samples[i]
  cat(paste("\nreading data:", sample, "\n"))

  seu.obj <- read.data(sample)
  seurat.list[[sample]] <- seu.obj
}

save(seurat.list, file = paste0(robjects.path,"/scRNA_SeuratObj_List.RData"))

merged.obj = merge(seurat.list[[1]], y = c(seurat.list[[2]],seurat.list[[3]],seurat.list[[4]],
                                           seurat.list[[5]],seurat.list[[6]],seurat.list[[7]],
                                           seurat.list[[8]],seurat.list[[9]],seurat.list[[10]],
                                           seurat.list[[11]],seurat.list[[12]]),
                   add.cell.ids = names(seurat.list))

merged.obj$orig.ident = paste0(sapply(strsplit(rownames(merged.obj@meta.data),"_"),"[",2),"_",
                               sapply(strsplit(rownames(merged.obj@meta.data),"_"),"[",3))

merged.obj[["percent.mt"]] = PercentageFeatureSet(merged.obj, pattern = "^mt-")
merged.obj[["percent.ribosomal"]] = PercentageFeatureSet(merged.obj, pattern = "^Rp[ls]")

save(merged.obj, file = paste0(robjects.path,"/scRNA_SeuratObj_Merged.RData"))

metadata = merged.obj@meta.data
save(metadata, file = paste0(robjects.path,"/scRNA_SeuratObj_Merged_Metadata.RData"))
```

## Quality control

### Mapping metrics

```{r mapping-met, eval = F}
load(paste0(robjects.path,"/scRNA_SeuratObj_Merged_Metadata.RData"))
samples = unique(metadata$orig.ident)

for(i in 1:length(samples)){
  metrics = read.csv(paste0(data.path,"/Sample_",samples[i],"/outs/metrics_summary.csv"))
  if(i == 1){
    metrics.df = metrics
  }else{
    metrics.df = rbind(metrics.df, metrics)
  }
}

rownames(metrics.df) = samples
write.csv(metrics.df, file = paste0(tables.path,"/Mapping_Metrics.csv"))
```

## Look at common metrics

```{r qc, eval = F}
load(paste0(robjects.path,"/scRNA_SeuratObj_Merged_Metadata.RData"))

metadata$log10GenesPerUMI = log10(metadata$nFeature_RNA) / log10(metadata$nCount_RNA)

metadata %>% 
  	ggplot(aes(x=orig.ident, fill=orig.ident)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells")

metadata.long = gather(metadata, type, value, 2:5)

metadata.long %>%
  ggplot(aes(x = orig.ident, y = value, fill = orig.ident)) + 
  geom_violin() + 
  # geom_point() +
  facet_wrap(vars(type), ncol = 2, scales = "free") 

# log scaled
metadata %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() +
  	theme_classic() +
    geom_vline(xintercept = 500) +
  	ylab("Cell density") +
    xlab("nCount_RNA (nUMI)")

metadata %>% 
  	ggplot(aes(color=orig.ident, x=nFeature_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() +
  	geom_vline(xintercept = c(250,9000)) + 
    xlab("nFeature_RNA (nGene)")

metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	# geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~orig.ident)

metadata %>% 
  	ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	# scale_x_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 25)

metadata %>%
  	ggplot(aes(x=log10GenesPerUMI, color = orig.ident, fill=orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)

metadata %>%
  ggplot(aes(x=log10GenesPerUMI, y=nFeature_RNA, color = log10(nCount_RNA))) +
  geom_point() + 
  facet_wrap(~orig.ident)

metadata %>%
  group_by(orig.ident) %>%
  summarise(Ncells = n(), nCount_RNA = mean(nCount_RNA), nFeature_RNA = mean(nFeature_RNA))

metadata %>%
  group_by(orig.ident) %>%
  summarise(Ncells = n(), nCount_RNA = median(nCount_RNA), nFeature_RNA = median(nFeature_RNA))
```

## Filtering

These thresholds for filtering will depend on the experiment.

```{r filt, eval = F}
metadata.sub = subset(metadata, (nCount_RNA >= 500) & 
                                (nFeature_RNA >= 250) &
                                (nFeature_RNA <=9000) &
                                (log10GenesPerUMI > 0.8) & 
                                (percent.mt < 25))
table(metadata$orig.ident)
table(metadata.sub$orig.ident)

metadata.sub %>%
  group_by(orig.ident) %>%
  summarise(Ncells = n(), nCount_RNA = mean(nCount_RNA), nFeature_RNA = mean(nFeature_RNA))

metadata.sub %>%
  group_by(orig.ident) %>%
  summarise(Ncells = n(), nCount_RNA = median(nCount_RNA), nFeature_RNA = median(nFeature_RNA))
```

## Quality control after filtering

```{r filt-qc, eval = F}
metadata = metadata.sub

metadata %>% 
  	ggplot(aes(x=orig.ident, fill=orig.ident)) + 
  	geom_bar() +
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells")

metadata.long = gather(metadata, type, value, 2:5)

metadata.long %>%
  ggplot(aes(x = orig.ident, y = value, fill = orig.ident)) + 
  geom_violin() + 
  # geom_point() +
  facet_wrap(vars(type), ncol = 2, scales = "free") 

# log scaled
metadata %>% 
  	ggplot(aes(color=orig.ident, x=nCount_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() +
  	theme_classic() +
    geom_vline(xintercept = 500) +
  	ylab("Cell density") +
    xlab("nCount_RNA (nUMI)")

metadata %>% 
  	ggplot(aes(color=orig.ident, x=nFeature_RNA, fill= orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	theme_classic() +
  	scale_x_log10() +
  	geom_vline(xintercept = c(250,9000)) +
  	ggtitle("nGenes per Cell")

metadata %>% 
  	ggplot(aes(x=nCount_RNA, y=nFeature_RNA, color=percent.mt)) + 
  	geom_point() + 
	scale_colour_gradient(low = "gray90", high = "black") +
  	stat_smooth(method=lm) +
  	scale_x_log10() +
  	scale_y_log10() +
  	theme_classic() +
  	# geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250) +
  	facet_wrap(~orig.ident)

metadata %>% 
  	ggplot(aes(color=orig.ident, x=percent.mt, fill=orig.ident)) + 
  	geom_density(alpha = 0.2) + 
  	# scale_x_log10() +
  	theme_classic() +
  	geom_vline(xintercept = 10)

metadata %>%
  	ggplot(aes(x=log10GenesPerUMI, color = orig.ident, fill=orig.ident)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)

metadata %>%
  ggplot(aes(x=log10GenesPerUMI, y=nFeature_RNA, color = log10(nCount_RNA))) +
  geom_point() + 
  facet_wrap(~orig.ident)
```

## Cell-level and gene-level filtering

These cell filters will change depending on experiment.

Additional cell filters:

- nGenes >= 250 & nGenes <=9000

- log10GenesPerUMI >= 0.8

- percent.mt < 25

Gene filters:

- Remove a gene if it is not expressed in at least 10 cells

```{r filt-cell, eval = F}
merged.obj$log10GenesPerUMI = log10(merged.obj$nFeature_RNA) / log10(merged.obj$nCount_RNA)

filtered_seurat = subset(x = merged.obj,
                         subset = 
                           (nFeature_RNA >= 250) & 
                           (nFeature_RNA <= 9000) &
                           (log10GenesPerUMI >= 0.8) & 
                           (percent.mt < 25))
filtered_seurat

save(filtered_seurat, file = paste0(robjects.path,"/scRNA_SeuratObj_Merged_CellFiltered.RData"))
```

```{r filt-gene, eval = F}
load(file = paste0(robjects.path,"/scRNA_SeuratObj_Merged_CellFiltered.RData"))
filtered_seurat[["RNA"]] = JoinLayers(filtered_seurat[["RNA"]])
filtered_seurat

counts = GetAssayData(object = filtered_seurat, assay = "RNA", layer = "counts")
keep_genes = Matrix::rowSums(counts) >= 10
filtered_counts = counts[keep_genes,]
dim(filtered_counts)

filtered_seurat2 = CreateSeuratObject(filtered_counts, meta.data = filtered_seurat@meta.data)
filtered_seurat2

filtered_seurat2[["RNA"]] = split(filtered_seurat2[["RNA"]], f = filtered_seurat2$orig.ident)
filtered_seurat2

save(filtered_seurat2, file = paste0(robjects.path,"/scRNA_SeuratObj_Merged_CellGeneFiltered.RData"))
```

## Doublets

If you want to remove doublets. I usually skip this step and opt for looking for doublet clusters after clustering and gene expression analysis.

Doublets should be estimated for each sample separately.

```{r doublets, eval = F}
filtered_seurat_list = SplitObject(filtered_seurat2, split.by = "orig.ident")

sce_list = lapply(filtered_seurat_list, function(x){
  y = as.SingleCellExperiment(x)
  return(y)
})

sce_dbl_list = lapply(sce_list, function(x){
  print(x)
  y = scDblFinder(x)
  return(y)
  }
)

save(sce_dbl_list, file = paste0(robjects.path,"/SCE_Doublets_List.RData"))

filtered_dbl_seurat_list = lapply(sce_dbl_list, function(x){
  y = as.Seurat(x, data = NULL)
  print(y)
  return(y)
})

save(filtered_dbl_seurat_list, file = paste0(robjects.path,"/scRNA_SeuratObj_List_CellGeneFiltered_DblRmv.RData"))
```

## Integration

### LogNormalize

```{r integration-log, eval = F}
load(file = paste0(robjects.path,"/scRNA_SeuratObj_Merged_CellGeneFiltered.RData"))

seu.list = SplitObject(filtered_seurat2, split.by = "orig.ident")
# run standard anlaysis workflow
seu.list = lapply(X = seu.list, FUN = function(x){
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, nfeatures = 5000)
})
#############################################################################################################
#############################################################################################################
features <- SelectIntegrationFeatures(object.list = seu.list, nfeatures = 5000)
#############################################################################################################
#############################################################################################################
seu.list <- lapply(X = seu.list, FUN = function(x){
  x <- ScaleData(x, features = features)
  x <- RunPCA(x, features = features)
})
#############################################################################################################
#############################################################################################################
anchors <- FindIntegrationAnchors(object.list = seu.list, anchor.features = features, reduction = "rpca")
#############################################################################################################
#############################################################################################################
int.obj <- IntegrateData(anchorset = anchors, normalization.method = "LogNormalize")
save(int.obj, file = paste0(robjects.path,"/scRNA_SeuratObj_LogNorm_IntegratedRPCA.RData"))
#############################################################################################################
#############################################################################################################
DefaultAssay(int.obj) <- "integrated"

# Run the standard workflow for visualization and clustering
int.obj <- ScaleData(int.obj, verbose = FALSE)
int.obj <- RunPCA(int.obj, reduction.name = "integrated.rpca")
int.obj <- RunUMAP(int.obj, reduction = "integrated.rpca", dims = 1:30, reduction.name = "umap.rpca")
int.obj <- FindNeighbors(int.obj, reduction = "integrated.rpca", dims = 1:30)
int.obj <- FindClusters(int.obj, resolution = 1, cluster.name = "rpca_clusters_res1.0")
int.obj
save(int.obj, file = paste0(robjects.path,"/scRNA_SeuratObj_LogNorm_IntegratedRPCA_Clustered_UMAP.RData"))
```

### SCTransform

```{r integration-sct, eval = F}
load(file = paste0(robjects.path,"/scRNA_SeuratObj_Merged_CellGeneFiltered.RData"))

seu.list <- SplitObject(filtered_seurat2, split.by = "orig.ident")

seu.list <- lapply(X = seu.list, FUN = function(x){
  x <- SCTransform(x)
})
#############################################################################################################
#############################################################################################################
features <- SelectIntegrationFeatures(object.list = seu.list, nfeatures = 5000)
#############################################################################################################
#############################################################################################################
seu.list <- PrepSCTIntegration(object.list = seu.list, anchor.features = features)
#############################################################################################################
#############################################################################################################
anchors <- FindIntegrationAnchors(object.list = seu.list, anchor.features = features, reduction = "rpca",
                                 normalization.method = "SCT")
#############################################################################################################
#############################################################################################################
int.obj <- IntegrateData(anchorset = anchors, normalization.method = "SCT")
save(int.obj, file = paste0(robjects.path,"/scRNA_SeuratObj_SCT_IntegratedRPCA.RData"))
#############################################################################################################
#############################################################################################################
DefaultAssay(int.obj) <- "integrated"
int.obj <- RunPCA(int.obj, reduction.name = "integrated.rpca")
int.obj <- RunUMAP(int.obj, reduction = "integrated.rpca", dims = 1:30, reduction.name = "umap.rpca")
int.obj <- FindNeighbors(int.obj, reduction = "integrated.rpca", dims = 1:30)
int.obj <- FindClusters(int.obj, resolution = 1, cluster.name = "rpca_clusters_res1.0")
int.obj
save(int.obj, file = paste0(robjects.path,"/scRNA_SeuratObj_SCT_IntegratedRPCA_Clustered_UMAP.RData"))
```

### Plotting

```{r plots, eval = F}
plots.path = "/path/to/plots"

pdf(file = paste0(plots.path,"/RPCA_Integrated_OrigIdent.pdf"), width = 12, height = 12)
DimPlot(int.obj, reduction = "integrated.rpca", group.by = "orig.ident")
dev.off()

pdf(file = paste0(plots.path,"/UMAP_Integrated_OrigIdent.pdf"), width = 12, height = 12)
DimPlot(int.obj, reduction = "umap.rpca", group.by = "orig.ident")
dev.off()

pdf(file = paste0(plots.path,"/UMAP_Integrated_Cluster.pdf"), width = 12, height = 12)
DimPlot(int.obj, reduction = "umap.rpca", group.by = "rpca_clusters_res1.0", label = T)
dev.off()

pdf(file = paste0(plots.path,"/UMAP_Integrated_OrigIdent_Cluster.pdf"), width = 16, height = 16)
DimPlot(int.obj, reduction = "umap.rpca", group.by = "rpca_clusters_res1.0", split.by = "orig.ident", label = T, ncol = 4)
dev.off()
```

## Identifying Marker Genes

Seurat recommend switching to RNA assay for amrker selection.

```{r markers, eval = F}
Idents(int.obj) <- "rpca_clusters_res.10"
DefaultAssay(int.obj) <- "RNA"
# int.obj[["RNA"]] <- JoinLayers(int.obj[["RNA"]])
int.obj <- NormalizeData(int.obj, assay = "RNA")
int.obj <- FindVariableFeatures(int.obj, assay = "RNA")
int.obj <- ScaleData(int.obj, assay = "RNA")

markers <- FindConservedMarkers(int.obj, grouping.var = "orig.ident")
```

