---
title: "scRNA-seq Pipeline"
author: "Emily Nissen"
date: "2024-02-07"
output: 
  github_document:
    html_preview: TRUE
    toc: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Pre-processing

See the "PreProcess" folder for scripts that can be downloaded based on the code in the following steps.

If the sequencing was done at the KUMC genomics core - they most likely used the NuGEN Ovation RRBS Methyl-Seq technology. These steps follow the recommendations found here: <https://github.com/nugentechnologies/NuMetRRBS>

## Convert BCL to FastQ files

```{r convert, eval = F}
runfolder <- "/path/to/run/folder/"

convert <- paste0("bcl2fastq --runfolder-dir ", runfolder, 
                  " --input-dir ", runfolder, "Data/Intensities/BaseCalls/ --output-dir ",
                  runfolder, "Unaligned/ --sample-sheet ", runfolder, 
                  "SampleSheet.csv --use-bases-mask Y*,I8Y*,Y* --minimum-trimmed-read-length 0 --mask-short-adapter-reads 0")

print(convert)
system(convert)
```

## Concatenate files, trim adaptors, and FastQC

```{r cat-trim-fastqc, eval = F}
args <- commandArgs(trailingOnly=TRUE)
group <- as.numeric(args[1])

dir <- "/path/to/run/folder/Unaligned/Project_/"
out <- "/path/to/out/folder/trimmedData/"
out2 <- "/path/to/out/folder/fastqc/"

samples <- c()
files <- paste0("Sample_",samples)
# split into however many groups you want
if(group == 1){
  samples = samples[]
  files = files[]
}else if(group == 2){
  samples = samples[]
  files = files[]
}

i=1
for(file in files){
  zcat <- paste0("zcat ", dir, file, "/*R1*.fastq.gz >> ", dir, file, "/", samples[i], "_R1.fastq")
  print(zcat)
  system(zcat)
  print(gsub("R1","R3",zcat))
  system(gsub("R1","R3",zcat))
  
  trim <- paste0("/kuhpc/work/biostat/e617n596/tools/TrimGalore-0.6.6/trim_galore --output_dir ", out, 
                " --paired -a AGATCGGAAGAGC -a2 AAATCAAAAAAAC ", 
                dir, file, "/", samples[i], "_R1.fastq ",
                dir, file, "/", samples[i], "_R3.fastq")
  print(trim)
  system(trim)
  
  fast <- paste0("fastqc -o ", out2, " -f fastq -t 8 ", out, samples[i], "_R1_val_1.fq")
  print(fast)
  system(fast)
  print(gsub("R1_val_1","R3_val_2",fast))
  system(gsub("R1_val_1","R3_val_2",fast))
  
  i = i+1
}
```

## Diversity Trim and Filtering

```{r}
args <- commandArgs(trailingOnly=TRUE)
group <- as.numeric(args[1])

dir <- "/path/to/working/dir/"
out.dir <- paste0(dir, "trimmedData/")

samples <- c()

if(group == 1){
  samples = samples[]
}else if(group == 2){
  samples = samples[]
}

for(sample in samples){
  trim <- paste0("python ", dir, "/trimRRBSdiversityAdaptCustomers.py -1 '", 
                out.dir, sample, "_R1_val_1.fq' -2 '",
                out.dir, sample, "_R3_val_2.fq'")
  print(trim)
  system(trim)
  
  move <- paste0("mv ", dir, sample,
                "_R*_trimmed.fq ", out.dir, "complete/")
  print(move)
  system(move)
}
```

## Align using Bismark

```{r}
args <- commandArgs(trailingOnly=TRUE)
group <- as.numeric(args[1])

dir <- "/path/to/working/dir/"
data.dir <- paste0(dir, "trimmedData/complete/")
out.dir <- paste0(dir, "bismarkAlignments/")

samples <- c()

if(group == 1){
  samples = samples[]
}else if(group == 2){
  samples = samples[]
}

for(sample in samples){
  bismark <- paste0("/kuhpc/work/biostat/e617n596/tools/Bismark-0.22.3/bismark --bowtie2 --path_to_bowtie2 /kuhpc/software/7/install/bowtie2/2.3.5.1/ --output_dir ", 
                    out.dir, " --multicore 8 --samtools_path /kuhpc/software/7/install/samtools/1.9/bin /kuhpc/work/biostat/e617n596/References/bismark/hg38 -1 ", 
                    dir, sample, "_R1_val_1_trimmed.fq -2 ", dir, sample, "_R3_val_2_trimmed.fq")
  
  print(bismark)
  systme(bismark)
}
```

